gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/visitor/ExecutionVisitor.java

```diff
@@ -329,9 +304,66 @@ public EntityFetcherResponse visit(NoOpNode noOpNode) {
     return new EntityFetcherResponse();
   }
 
-  @VisibleForTesting
-  boolean isSingleSourceAndSame(Set<String> firstSource, Set<String> secondSource) {
-    return firstSource.size() == 1 &&
-        firstSource.equals(secondSource);
+  @Override
+  public EntityFetcherResponse visit(SelectionAndFilterNode selectionAndFilterNode) {
+    List<EntityFetcherResponse> resultMapList = new ArrayList<>();
+    EntitiesRequest request =
+        EntitiesRequest.newBuilder(executionContext.getEntitiesRequest())
+            .setOffset(selectionAndFilterNode.getOffset())
+            .setLimit(selectionAndFilterNode.getLimit())
+            .build();
+    EntitiesRequestContext context =
+        new EntitiesRequestContext(
+            executionContext.getTenantId(),
+            request.getStartTimeMillis(),
+            request.getEndTimeMillis(),
+            request.getEntityType(),
+            executionContext.getRequestHeaders());
+
+    String source = selectionAndFilterNode.getSource();
+    if (source.equals("QS")) {
```

Comment:
Such source based checks in the ExecutionVisitor is troubling me.
I did a line by line comparison of the getEntities, getAggregatedMetrics and getEntitiesAndAggregatedMetrics in QueryServiceEntityFetcher and all of them are exactly the same apart from the below which i think can just be conditionally set.
```
 builder.setLimit(entitiesRequest.getLimit());
    builder.setOffset(entitiesRequest.getOffset());
    builder.addAllOrderBy(
         QueryAndGatewayDtoConverter.convertToQueryOrderByExpressions(entitiesRequest.getOrderByList()));

```

So if we can just wrap up all of those into a single method getEntities, then this custom checks are also not needed. We can just call getEntities here too.

Reply:
Different PR once we add some more tests.